build-databroker.sh

```diff
@@ -0,0 +1,160 @@
+if [ -z "$KUKSA_DATABROKER_FEATURES" ]; then
+    # If not set, assign a default value
+    KUKSA_DATABROKER_FEATURES="databroker/default"
+fi
+
+SBOM=0
+# Check whether to build SBOM
+if [ ! -z "$KUKSA_DATABROKER_SBOM" ]; then
+    # If set, check whether it is "y"
+    if [[ $KUKSA_DATABROKER_SBOM =~ ^[Yy](es)?$ || $KUKSA_DATABROKER_SBOM =~ ^[Tt](rue)?$ ]]; then
+        SBOM=1
+    fi
+fi
+
+if [[ $SBOM -eq 1 ]]; then
+    echo "Will create SBOM"
+else
+    echo "Will not create SBOM"
+fi
+
+echo  "Building with features: $KUKSA_DATABROKER_FEATURES"
+
+
+
+
+# Builds for a given target and collects data to be distirbuted in docker. Needs
+# Rust target triplett (i.e. x86_64-unknown-linux-musl) and the corresponding docker
+# architecture (i.e. amd64) as input
+function build_target() {
+    target_rust=$1
+    target_docker=$2
+
+    echo "Building databroker for target $target_rust"
+    cross build --target $target_rust --features $KUKSA_DATABROKER_FEATURES --bin databroker --release
+
+    echo "Prepare $target_docker dist folder"
+    mkdir ./dist/$target_docker
+    cp ./target/$target_rust/release/databroker ./dist/$target_docker
+
+    if [[ $SBOM -eq 1 ]]; then
+        echo "Create $target_rust SBOM"
+        cargo cyclonedx -v -f json --describe binaries --spec-version 1.4 --target $target_rust --manifest-path ./Cargo.toml
+        cp ./databroker/databroker_bin.cdx.json ./dist/$target_docker/sbom.json
+        rm -rf ./dist/$target_docker/thirdparty-licenses || true
+        collectlicensefiles ./databroker/databroker_bin.cdx.json ./dist/$target_docker/thirdparty-licenses --curation ./licensecuration.yaml
+    fi
+
+    # We need to clean this folder in target, otherwise we get weird side
+    # effects building the aarch image, complaining libc crate can not find
+    # GLIBC, i.e
+    #   Compiling libc v0.2.149
+    #error: failed to run custom build command for `libc v0.2.149`
+    #
+    #Caused by:
+    #  process didn't exit successfully: `/target/release/build/libc-2dd22ab6b5fb9fd2/#build-script-build` (exit status: 1)
+    #  --- stderr
+    #  /target/release/build/libc-2dd22ab6b5fb9fd2/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.29' not found (required by /target/release/build/libc-2dd22ab6b5fb9fd2/build-script-build)
+    #
+    # It seems cross/cargo is reusing something from previous builds it shouldn't.
+    # the finished artifact resides in ../target/x86_64-unknown-linux-musl/release
+    # so deleting the temporary files in target/releae is no problem
+    echo "Cleaning up...."
+    rm -rf ./target/release
+}
+
+
+# Check valid platforms
+for platform in "$@"
+do
+    if [ ! -f ${TARGET_MAP}/$platform ]; then
+        echo "Invalid platform \"$platform\""
+        echo "Supported platforms:"
+        echo "$(ls ${TARGET_MAP})"
+        rm -rf ${TARGET_MAP}
+        exit 1
+    fi
+done
+
+
+# Starting a fresh build
+echo "Cargo clean, to start fresh..."
+cargo clean
```

Comment:
This breaks caching.

I don't think it's a good default for a script to force removal of all build caches. That is better controlled outside the script or added as an optional argument if deemed useful. Since each platform is built in a separate runner, we do not need to do a cargo clean in CI anyway (that is why it was added, right?). For one thing, it removes the cached install of `cargo cross`, which makes the build slower. It also removes the cache of dependencies, further making the build slower.

Reply:
This is out now. We are still deleting target/release in the build step (see comment in the script), so this might not save mich time currently

Reply:
Ok I checked, the `rm -rf ./target/release` is still necessary, becasue

```
rm -rf target
rm -rf dist
./build-databroker.sh amd64
./build-databroker.sh riscv64
```

still yields

```
 Compiling bytes v1.6.0
error: failed to run custom build command for `libc v0.2.153`

Caused by:
  process didn't exit successfully: `/target/release/build/libc-71ad6c3e8bc9cca0/build-script-build` (exit status: 1)
  --- stderr
  /target/release/build/libc-71ad6c3e8bc9cca0/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.29' not found (required by /target/release/build/libc-71ad6c3e8bc9cca0/build-script-build)
  /target/release/build/libc-71ad6c3e8bc9cca0/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /target/release/build/libc-71ad6c3e8bc9cca0/build-script-build)
warning: build failed, waiting for other jobs to finish...
```

Reply:
But that is not the workflow when running in CI, so why should the CI build time be increased because of that..?

Reply:
target is in cache, so I assume the exact same can happen in CI, it just depends on the order of builds and what state of cache gets reinstantiated

Reply:
Nope, since the cache key contains the platform, a build directory that was used for building a different platform / arch will never be pulled.

```yaml
key: databroker-release-${{ matrix.platform.name }}-${{ hashFiles('**/Cargo.lock') }}
```

Reply:
But then `rm -rf ./target/release` does not have a negative impact on CI-builds, and is needed at least for local builds as shown by Sebastian, so no problem from a performance perspective to keep it, or?

Reply:
> But then `rm -rf ./target/release` does not have a negative impact on CI-builds

What do you mean?

If you `rm -rf` the cache before building, the CI will obviously not be able to use what was in the cache.. that will impact the performance.

Reply:
My mistake.

For my needs to be able to build locally it would be no problem to have an opt-in for clean, possibly something like below causing a clean of target/release before each target being built.

`./build-databroker.sh clean arm64 amd64`

Reply:
I think I found a way now to safely cache, that also works locally. You can tell cargo to use  different directories for target, I split now depending on architecture.